ethcore/src/miner/miner.rs

```diff
@@ -729,27 +740,43 @@ impl MinerService for Miner {
 	}
 
 	fn update_sealing(&self, chain: &MiningBlockChainClient) {
-		if self.sealing_enabled.load(atomic::Ordering::Relaxed) {
-			let current_no = chain.chain_info().best_block_number;
-			let has_local_transactions = self.transaction_queue.lock().has_local_pending_transactions();
-			let last_request = *self.sealing_block_last_request.lock();
-			let should_disable_sealing = !self.forced_sealing()
-				&& !has_local_transactions
-				&& current_no > last_request
-				&& current_no - last_request > SEALING_TIMEOUT_IN_BLOCKS;
-
-			if should_disable_sealing {
-				trace!(target: "miner", "Miner sleeping (current {}, last {})", current_no, last_request);
-				self.sealing_enabled.store(false, atomic::Ordering::Relaxed);
-				self.sealing_work.lock().reset();
+		trace!(target: "miner", "update_sealing");
+		let requires_reseal = {
+			let mut sealing_work = self.sealing_work.lock();
+			if sealing_work.enabled {
+				trace!(target: "miner", "update_sealing: sealing enabled");
+				let current_no = chain.chain_info().best_block_number;
+				let has_local_transactions = self.transaction_queue.lock().has_local_pending_transactions();
+				let last_request = *self.sealing_block_last_request.lock();
+				let should_disable_sealing = !self.forced_sealing()
+					&& !has_local_transactions
+					&& current_no > last_request
+					&& current_no - last_request > SEALING_TIMEOUT_IN_BLOCKS;
+
+				trace!(target: "miner", "update_sealing: should_disable_sealing={}; current_no={}, last_request={}", should_disable_sealing, current_no, last_request);
+
+				if should_disable_sealing {
+					trace!(target: "miner", "Miner sleeping (current {}, last {})", current_no, last_request);
+					sealing_work.enabled = false;
+					sealing_work.queue.reset();
+					false
+				} else {
+					// sealing enabled and we don't want to sleep.
+					*self.next_allowed_reseal.lock() = Instant::now() + self.options.reseal_min_period;
+					true
+				}
 			} else {
-				*self.next_allowed_reseal.lock() = Instant::now() + self.options.reseal_min_period;
-				// --------------------------------------------------------------------------
-				// | NOTE Code below requires transaction_queue and sealing_work locks.     |
-				// | Make sure to release the locks before calling that method.             |
-				// --------------------------------------------------------------------------
-				self.prepare_sealing(chain);
+				// sealing is disabled.
+				false
 			}
+		};
+		
+		if requires_reseal {
+			// --------------------------------------------------------------------------
+			// | NOTE Code below requires transaction_queue and sealing_work locks.     |
```

Comment:
could perhaps do to be made clear that the function doesn't actually need it to be locked, but rather needs it to be unlocked, since it needs to lock it.